STP-3522: Clarify external system info in integration (#89)
Cray-HPE/docs-sat: docs/external_system.md

1. Clone the SAT source code.

   **Note:** To use SAT version 3.19, this example clones the `release/3.19`
   branch of `Cray-HPE/sat`. However, for better clarity, these instructions
   include steps that apply only to versions newer than 3.19. Specifically, the
   instructions include references to the `csm-api-client` package, which was

>    not a dependency of SAT in version 3.19.

Author: I'll update to the following in the next commit:

```
To use SAT version 3.21, this example clones the `release/3.21 branch of 
`Cray-HPE/sat`.
```

   ```screen

>    (venv) $ git clone --branch=release/3.19 https://github.com/Cray-HPE/sat.git

Author: Updated in next commit

   endpoint = "http://s3.system-name.site-domain:8080"
   ```

1. Edit the SAT configuration file to specify the Keycloak user which will be
   accessing the REST API.

   ```toml
   [api_gateway]
   username = "user"
   ```

1. Run `sat auth`. Enter your password when prompted.

   The admin account used to authenticate with `sat auth` must be enabled in
   Keycloak and must have its *assigned role* set to *admin*. For more
   information on editing *Role Mappings*, see *Create Internal User Accounts
   in the Keycloak Shasta Realm* in the
   [Cray System Management Documentation](https://cray-hpe.github.io/docs-csm/).
   For more information on authentication types and authentication credentials,
   see [SAT Command Authentication](introduction.md#sat-command-authentication).

   ```screen
   $ sat auth

>    Password for crayadmin:

Reviewer: You could update this example to use the same username as in the example from the previous step.

```suggestion
   Password for user:
```

Author: Updated in next commit

   Succeeded!
   ```

> 1. Ensure the files are readable only by `root`.

Reviewer: Minor change needed here:

```suggestion
1. Ensure the files are readable only by the current user.
```

Author: Updated in next commit

    ```screen
    $ touch ~/.config/sat/s3_access_key \
        ~/.config/sat/s3_secret_key
    ```

    ```screen
    $ chmod 600 ~/.config/sat/s3_access_key \
        ~/.config/sat/s3_secret_key
    ```

1. Write the credentials to local files using `kubectl`.

>   Generate S3 credentials and write them to a local file so the SAT user can

Reviewer: All the lines under this step need to be indented by one more space to properly appear as a part of this step when rendered.

Author: Thanks for catching this! Updated in next commit.

>   access S3 storage. In order to use the SAT S3 bucket, the system
>   administrator must generate the S3 access key and secret keys and write them

Reviewer: Since this is happening on an external system, "user" is more appropriate than "system administrator".

I also think "generate" is a little misleading. This step really just gets these values from a Kubernetes secret and stores them in files. I know that "generate s3 credentials" terminology is used elsewhere, so feel free to defer that piece of feedback if you prefer.

```suggestion
  access S3 storage. In order to use the SAT S3 bucket, the user
  must obtain the S3 access key and secret keys and write them
```

Author: Updated in next commit.

  to a local file.

>   This must be done on every Kubernetes master node where SAT commands are run.

Reviewer: Drop this line since this is happening on an external system and not a Kubernetes master node.

```suggestion
```

Author: Updated in next commit.

> ## Authenticate SAT Commands

Reviewer: I'm looking at this more closely now, and I'm realizing that most of the after this point needs to be removed or  modified in order to make sense in the context of configuring SAT on an external system. In the previous steps in the section "Install and Configure SAT", we have already:

* Run `sat init` to generate a SAT config file (Step 11)
* Modified the configuration file to set the API gateway hostname and the s3 endpoint (Step 13)
* Modified the username in the SAT config file (Step 14)

All that really remains to do here is:

* Run `sat auth`. This is just step 4 of the "Authenticate SAT Commands" section
* Run steps 1 and 2 of "Generate SAT S3 Credentials" section below. This works because we set up our Kubernetes config file in steps 7-10 of the "Install and Configure SAT" procedure.

What do you think about adding these remaining things as additional steps in this "Install and Configure SAT" procedure? The reason I suggest doing it this way instead of preserving the two following sections is that we have already run `sat init`, and that's really integral to some of the other steps in the section, and we've also already set the username in the SAT config file.

Author: Sounds like a great idea! I'll update in the next commit.

## Authenticate SAT Commands

Initially, as part of the installation and configuration, SAT authentication is
set up so SAT commands can be used in later steps of the install process. The
admin account used to authenticate with `sat auth` must be enabled in Keycloak
and must have its *assigned role* set to *admin*. For more information on
editing *Role Mappings*, see *Create Internal User Accounts in the Keycloak
Shasta Realm* in the [Cray System Management Documentation](https://cray-hpe.github.io/docs-csm/).
For more information on authentication types and authentication credentials,
see [SAT Command Authentication](introduction.md#sat-command-authentication).

### Prerequisites

- The `sat` CLI has been installed following [Install and Configure

>   SAT](#install-and-configure-sat).

Reviewer: I would remove this prerequisite here. This is an alternate external-system install procedure, and it does not depend on installing and configuring SAT as described in the linked document.

Author: Removed in next commit

### Procedure

The following is the procedure to globally configure the username used by SAT
and authenticate to the API gateway.

1. Generate a default SAT configuration file, if one does not exist.

   ```screen
   $ sat init

>    Configuration file "/root/.config/sat/sat.toml" generated.

Reviewer: This line will have a different path, depending on the user's home directory on the system. E.g., on my macbook, it will be something like:

```
Configuration file "/Users/haasken/.config/sat/sat.toml" generated.
```

Perhaps we could do something like this, which isn't exactly as it appears, but should get the point across:

```suggestion
Configuration file "$HOME/.config/sat/sat.toml" generated.
```

Author: Updated in next commit

   ```

   **Note:** If the config file already exists, it will print out the following
   error.

   ```screen

>    ERROR: Configuration file "/root/.config/sat/sat.toml" already exists.
>    Not generating configuration file.

Reviewer: Similarly, the path here would need to change. I also noticed that this is a warning now, and the "Not generating configuration file" should be on the same line:

```suggestion
   WARNING: Configuration file "$HOME/.config/sat/sat.toml" already exists. Not generating configuration file.
```

   ```

1. Edit `~/.config/sat/sat.toml` and set the username option in the `api_gateway`
   section of the config file.

   ```screen
   username = "crayadmin"
   ```

1. Run `sat auth`. Enter your password when prompted.

   ```screen
   $ sat auth
   Password for crayadmin:
   Succeeded!
   ```

1. Other `sat` commands are now authenticated to make requests to the API gateway.

   ```screen
   $ sat status
   ```

## Generate SAT S3 Credentials

Generate S3 credentials and write them to a local file so the SAT user can
access S3 storage. In order to use the SAT S3 bucket, the System Administrator
must generate the S3 access key and secret keys and write them to a local file.
This must be done on every Kubernetes master node where SAT commands are run.

SAT uses S3 storage for several purposes, most importantly to store the
site-specific information set with `sat setrev`.

> ### Prerequisites

Author: Removed in next commit

1. Ensure the files are readable only by `root`.

    ```screen

>     $ touch /root/.config/sat/s3_access_key \
>         /root/.config/sat/s3_secret_key

Reviewer: These commands need to use the user's home directory. So this would work:

```suggestion
    $ touch ~/.config/sat/s3_access_key \
        ~/.config/sat/s3_secret_key
```

Author: Updated in next commit

    ```

    ```screen

>     $ chmod 600 /root/.config/sat/s3_access_key \
>         /root/.config/sat/s3_secret_key

Reviewer: ```suggestion
    $ chmod 600 ~/.config/sat/s3_access_key \
        ~/.config/sat/s3_secret_key
```

Author: Updated in next commit

    ```

1. Write the credentials to local files using `kubectl`.

   ```screen
   $ kubectl get secret sat-s3-credentials -o json -o \
       jsonpath='{.data.access_key}' | base64 -d > \

>        /root/.config/sat/s3_access_key

Reviewer: ```suggestion
       ~/.config/sat/s3_access_key
```

Author: Updated in next commit

   ```

   ```screen
   $ kubectl get secret sat-s3-credentials -o json -o \
       jsonpath='{.data.secret_key}' | base64 -d > \

>        /root/.config/sat/s3_secret_key

Author: Updated in next commit

      character, the SAT configuration will take the default value – `"https://rgw-vip.nmn"`.

      ```screen
      $ grep endpoint ~/.config/sat/sat.toml
      # endpoint = "https://rgw-vip.nmn"
      ```

   1. Get the `sat-s3-credentials` secret's endpoint value.

      ```screen
      $ kubectl get secret sat-s3-credentials -o json -o \
          jsonpath='{.data.s3_endpoint}' | base64 -d | xargs
      https://rgw-vip.nmn

>       ```

Author: Updated in next commit
